presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -252,8 +253,15 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
             return -round(-num, decimals);
         }
 
-        double factor = Math.pow(10, decimals);
-        return Math.floor(num * factor + 0.5) / factor;
+        double log10 = Math.log10(num);
+        int alldecimals = 0;
+        if (log10 < 0.0) {
+            alldecimals = (int) decimals;
+        }
+        else {
+            alldecimals = (int) Math.ceil(log10) + (int) decimals;
+        }
+        return new BigDecimal(num).round(new MathContext(alldecimals)).doubleValue();
```

Comment:
This is going to be too expensive. We need to find a way to fix this without allocating a BigDecimal object